enterprise/jaas-security/src/main/java/ClientLoginModule.java

```diff
@@ -0,0 +1,198 @@
+ * <p>
+ * Obviously you would NEVER store passwords in clear text.
+ *
+ */
+public class ClientLoginModule implements LoginModule  {
+
+    private final ILogger logger = Logger.getLogger(getClass().getName());
+    private Credentials credentials;
+    private Subject subject;
+    private CallbackHandler callbackHandler;
+    private Map sharedState;
+    private Map options;
+
+    // The Group that the user is authorised for.
+    private UserGroupCredentials userGroupCredentials;
+
+    // Usernames and Password are stored here in clear text, obviously NEVER do this.
+    private static Map<String,String> allowedUsersMap = new HashMap<String,String>();
+
+    // This map represents the userAssignedGroup(s) that a user would belong to, this would be the result from your LDAP store.
+    private static Map<String,String> userGroups = new HashMap<String,String>();
+
+    static{
+
+        allowedUsersMap.put("david","password1");
+        allowedUsersMap.put("chris","password2");
+
+        userGroups.put("david","adminGroup");
+        userGroups.put("chris","readOnlyGroup");
+
+    }
+
+    public void initialize(Subject subject, CallbackHandler callbackHandler, Map<String, ?> sharedState, Map<String, ?> options) {
+        this.subject = subject;
+        this.callbackHandler = callbackHandler;
+        this.sharedState = sharedState;
+        this.options = options;
+    }
+
+    /**
+     * Login is called when this module is executed.
+     *
+     * @return is login successful
+     * @throws LoginException
+     */
+    public boolean login() throws LoginException {
+        boolean loginOk = false;
+
+        final CredentialsCallback cb = new CredentialsCallback();
+        try {
+            callbackHandler.handle(new Callback[]{cb});
+            credentials = cb.getCredentials();
+        } catch (Exception e) {
+            throw new LoginException(e.getClass().getName() + ":" + e.getMessage());
+        }
+
+        if(credentials == null) {
+            logger.log(Level.WARNING, "Credentials could not be retrieved!");
+            return false;
+        }
+        logger.log(Level.INFO, "Authenticating " + SecurityUtil.getCredentialsFullName(credentials));
+
+        if (credentials instanceof UsernamePasswordCredentials){
+            loginOk = doLoginCheck((UsernamePasswordCredentials) credentials);
+        }
+
+        return loginOk;
+    }
+
+    private boolean doLoginCheck(UsernamePasswordCredentials credentials) {
+
+        String username = credentials.getUsername();
+        String password = allowedUsersMap.get(username);
+        boolean loginCheckOk = false;
+
+        if (password != null){
+            if(password.equals(credentials.getPassword())){
+                String userGroup = userGroups.get(username);
+                if (userGroup != null){
+                    userGroupCredentials = new UserGroupCredentials(credentials.getEndpoint(),userGroup);
```

Comment:
We don't add the authentication name part of the information into the Subject. Nonetheless, it's not a problem for this code sample.